File: pm_core_1/src/infrastructure/store/adapter/priority_adapter.go
/*
Author: QuanTuanHuy
Description: Part of Serp Project
*/

package adapter

import (
	"context"
	"fmt"

	"github.com/serp/pm-core/src/core/domain/entity"
	"github.com/serp/pm-core/src/core/port/store"
	"github.com/serp/pm-core/src/infrastructure/store/mapper"
	"github.com/serp/pm-core/src/infrastructure/store/model"
	"gorm.io/gorm"
)

type PriorityAdapter struct {
	db     *gorm.DB
	mapper *mapper.PriorityMapper
}

func NewPriorityAdapter(db *gorm.DB) store.IPriorityStore {
	return &PriorityAdapter{
		db:     db,
		mapper: mapper.NewPriorityMapper(),
	}
}

func (p *PriorityAdapter) CreatePriority(ctx context.Context, tx *gorm.DB, priority *entity.PriorityEntity) (*entity.PriorityEntity, error) {
	db := p.getDB(tx)
	priorityModel := p.mapper.ToModel(priority)
	if err := db.WithContext(ctx).Create(priorityModel).Error; err != nil {
		return nil, fmt.Errorf("failed to create priority: %w", err)
	}
	return p.mapper.ToEntity(priorityModel), nil
}

func (p *PriorityAdapter) ExistsByName(ctx context.Context, tenantID int64, name string) (bool, error) {
	var count int64
	if err := p.db.WithContext(ctx).Model(&model.PriorityModel{}).
		Where("tenant_id = ? AND name = ? AND deleted_at IS NULL", tenantID, name).
		Count(&count).Error; err != nil {
		return false, fmt.Errorf("failed to check priority existence: %w", err)
	}
	return count > 0, nil
}

func (p *PriorityAdapter) GetPriorityByID(ctx context.Context, ID int64, tenantID int64) (*entity.PriorityEntity, error) {
	var priorityModel model.PriorityModel
	if err := p.db.WithContext(ctx).Model(&model.PriorityModel{}).
		Where("id = ? AND tenant_id = ? AND deleted_at IS NULL", ID, tenantID).
		First(&priorityModel).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, nil
		}
		return nil, fmt.Errorf("failed to get priority by id: %w", err)
	}
	return p.mapper.ToEntity(&priorityModel), nil
}

func (p *PriorityAdapter) ListPriorities(ctx context.Context, tenantID int64) ([]*entity.PriorityEntity, error) {
	var priorityModels []*model.PriorityModel
	if err := p.db.WithContext(ctx).Model(&model.PriorityModel{}).
		Where("tenant_id = ? AND deleted_at IS NULL", tenantID).
		Find(&priorityModels).Error; err != nil {
		return nil, fmt.Errorf("failed to list priorities: %w", err)
	}
	return p.mapper.ToEntities(priorityModels), nil
}

func (p *PriorityAdapter) SoftDeletePriority(ctx context.Context, tx *gorm.DB, ID int64, tenantID int64) error {
	db := p.getDB(tx)
	if err := db.WithContext(ctx).
		Model(&model.PriorityModel{}).
		Where("id = ? AND tenant_id = ? AND deleted_at IS NULL", ID, tenantID).
		Update("deleted_at", gorm.Expr("CURRENT_TIMESTAMP")).Error; err != nil {
		return fmt.Errorf("failed to soft delete priority: %w", err)
	}
	return nil
}

func (p *PriorityAdapter) UpdatePriority(ctx context.Context, tx *gorm.DB, priority *entity.PriorityEntity) error {
	db := p.getDB(tx)
	priorityModel := p.mapper.ToModel(priority)
	if err := db.WithContext(ctx).Save(priorityModel).Error; err != nil {
		return fmt.Errorf("failed to update priority: %w", err)
	}
	return nil
}

func (p *PriorityAdapter) getDB(tx *gorm.DB) *gorm.DB {
	if tx != nil {
		return tx
	}
	return p.db
}

File: pm_core/src/main/java/serp/project/pmcore/infrastructure/store/adapter/PriorityAdapter.java
/**
 * Author: QuanTuanHuy
 * Description: Part of Serp Project
 */

package serp.project.pmcore.infrastructure.store.adapter;

import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Component;

import serp.project.pmcore.domain.priority.entity.PriorityEntity;
import serp.project.pmcore.domain.priority.port.IPriorityPort;
import serp.project.pmcore.domain.priority.query.PriorityListCriteria;
import serp.project.pmcore.domain.shared.pagination.PageResult;
import serp.project.pmcore.infrastructure.store.mapper.PriorityMapper;
import serp.project.pmcore.infrastructure.store.model.PriorityModel;
import serp.project.pmcore.infrastructure.store.repository.IPriorityRepository;
import serp.project.pmcore.infrastructure.store.support.PageableUtils;

import java.util.List;
import java.util.Optional;

@Component
@RequiredArgsConstructor
public class PriorityAdapter implements IPriorityPort {

    private final IPriorityRepository priorityRepository;
    private final PriorityMapper priorityMapper;

    @Override
    public PriorityEntity createPriority(PriorityEntity priority) {
        return priorityMapper.toEntity(priorityRepository.save(priorityMapper.toModel(priority)));
    }

    @Override
    public Optional<PriorityEntity> getPriorityById(Long id, Long tenantId) {
        return priorityRepository.findByIdAndTenantId(id, tenantId)
                .map(priorityMapper::toEntity);
    }

    @Override
    public Optional<PriorityEntity> getPriorityByIdIncludingSystem(Long id, Long tenantId) {
        return priorityRepository.findByIdAndTenantIdOrSystemTenant(id, tenantId)
                .map(priorityMapper::toEntity);
    }

    @Override
    public Optional<PriorityEntity> getPriorityByPriorityKey(Long tenantId, String priorityKey) {
        return priorityRepository.findFirstByTenantIdAndPriorityKeyOrderByIdAsc(tenantId, priorityKey)
                .map(priorityMapper::toEntity);
    }

    @Override
    public List<PriorityEntity> listPriorities(Long tenantId) {
        return priorityMapper.toEntities(priorityRepository.findAllByTenantIdOrderBySequenceAsc(tenantId));
    }

    @Override
    public List<PriorityEntity> listPrioritiesIncludingSystem(Long tenantId) {
        return priorityMapper.toEntities(priorityRepository.findAllByTenantIdOrSystemTenant(tenantId));
    }

    @Override
    public List<PriorityEntity> getPrioritiesByIdsIncludingSystem(List<Long> priorityIds, Long tenantId) {
        if (priorityIds == null || priorityIds.isEmpty()) {
            return List.of();
        }
        return priorityMapper.toEntities(
                priorityRepository.findAllByIdInAndTenantIdOrSystemTenant(priorityIds, tenantId)
        );
    }

    @Override
    public PageResult<PriorityEntity> listPrioritiesIncludingSystem(Long tenantId, PriorityListCriteria criteria) {
        Pageable pageable = PageableUtils.of(criteria, resolveSort(criteria));
        Page<PriorityModel> result = priorityRepository.findAllVisibleWithFilters(
                tenantId,
                criteria.getSearchPatternLower(),
                criteria.getIsSystem(),
                pageable
        );
        return new PageResult<>(priorityMapper.toEntities(result.getContent()), result.getTotalElements());
    }

    @Override
    public void updatePriority(PriorityEntity priority) {
        priorityRepository.save(priorityMapper.toModel(priority));
    }

    @Override
    public void deletePriority(Long id, Long tenantId) {
        priorityRepository.deleteByIdAndTenantId(id, tenantId);
    }

    @Override
    public boolean existsByName(Long tenantId, String name) {
        return priorityRepository.existsByTenantIdAndNameIgnoreCase(tenantId, name);
    }

    private Sort resolveSort(PriorityListCriteria criteria) {
        String sortBy = criteria.getSortBy().toLowerCase();
        Sort.Direction direction = PageableUtils.resolveDirection(criteria.getSortDirection());

        return switch (sortBy) {
            case "sequence" -> Sort.by(
                    new Sort.Order(direction, "sequence"),
                    new Sort.Order(direction, "name"),
                    new Sort.Order(direction, "id")
            );
            case "name" -> Sort.by(
                    new Sort.Order(direction, "name"),
                    new Sort.Order(direction, "id")
            );
            case "created_at" -> Sort.by(
                    new Sort.Order(direction, "createdAt"),
                    new Sort.Order(direction, "id")
            );
            default -> throw new IllegalArgumentException("sortBy must be one of sequence, name, created_at");
        };
    }
}
